server/routes/utils/adHelpers.ts

```diff
@@ -253,3 +253,93 @@ export const getErrorMessage = (err: any) => {
     ? get(err, 'body.error.reason')
     : get(err, 'message');
 };
+
+// Currently: detector w/ detection date range is considered a 'historical' detector
+export const getHistoricalDetectors = (detectors: Detector[]) => {
+  return detectors.filter(
+    (detector) => detector.detectionDateRange !== undefined
+  );
+};
+
+// Currently: detector w/ no detection date range is considered a 'realtime' detector
+export const getRealtimeDetectors = (detectors: Detector[]) => {
+  return detectors.filter(
+    (detector) => detector.detectionDateRange === undefined
+  );
+};
+
+export const getDetectorTasks = (detectorTaskResponses: any[]) => {
+  const detectorToTaskMap = {} as { [key: string]: any };
+  detectorTaskResponses.forEach((response) => {
+    const detectorId = get(response, '_id', '');
+    const detectorTask = get(response, 'anomaly_detection_task', null);
+    if (detectorTask !== null) {
+      detectorToTaskMap[detectorId] = detectorTask;
+    }
+  });
+  return detectorToTaskMap;
+};
+
+export const getDetectorResults = (detectorResultsResponses: any[]) => {
+  const detectorToResultsMap = {} as { [key: string]: any };
+  detectorResultsResponses.forEach((response) => {
+    const detectorId = get(response, 'hits.hits.0._source.detector_id', null);
+    if (detectorId !== null) {
+      detectorToResultsMap[detectorId] = response;
+    }
+  });
+  return detectorToResultsMap;
+};
+
+// Append task-related info - task state & anomaly results of the task.
+// If there is no related task info for a detector: set to default values of DISABLED state and 0 anomalies
+export const appendTaskInfo = (
+  detectorMap: { [key: string]: any },
+  detectorToTaskMap: { [key: string]: any },
+  detectorToResultsMap: { [key: string]: any }
+) => {
+  const detectorMapWithTaskInfo = {} as { [key: string]: Detector };
+  const detectorsWithTasks = Object.keys(detectorToTaskMap);
+  Object.keys(detectorMap).forEach((detectorId, index) => {
+    if (!detectorsWithTasks.includes(detectorId)) {
+      detectorMapWithTaskInfo[detectorId] = {
+        ...detectorMap[detectorId],
+        curState: DETECTOR_STATE.DISABLED,
+        totalAnomalies: 0,
+      };
+    } else {
+      const task = detectorToTaskMap[detectorId];
+      const state = getHistoricalDetectorState(task);
+      const totalAnomalies = get(
+        detectorToResultsMap[detectorId],
+        'hits.total.value',
+        0
+      );
+      detectorMapWithTaskInfo[detectorId] = {
+        ...detectorMap[detectorId],
+        curState: state,
+        totalAnomalies: totalAnomalies,
+      };
+    }
+  });
+  return detectorMapWithTaskInfo;
+};
+
+// Four checks/transformations need to be made here:
+// (1) set to DISABLED if there is no existing task for this detector
+// (2) set to UNEXPECTED_FAILURE if the task is in a FAILED state to stay consistent
+// (3) set to INIT if the task is in a CREATED state
+// (4) set to DISABLED if the task is in a STOPPED state
```

Comment:
Does not cover all stats , check all task states https://github.com/opendistro-for-elasticsearch/anomaly-detection/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/ad/model/ADTaskState.java#L19.

INIT, RUNNING and FINISHED can be used directly like realtime detector?

Reply:
So the 4 checks are just to transform any of the returned historical detector states to match up with the states recognized by the frontend (the `DETECTOR_STATE` enum [here](server/utils/constants.ts)). And yes, if the state is `INIT`/`RUNNING`/`FINISHED`, since those are all recognized frontend states, no transformation is needed.

I decided to combine the real-time and historical states into one enum because of the end goal of having both detector types be a more singular data model.

Please let me know if you have any other suggestions here regarding the changes here. My reasoning for not exposing the `CREATED` state, and instead encapsulating it as part of the `INIT` state, was just to simplify the frontend changes and not add too many new terms and states for the user to have to understand.

Reply:
Sounds good. `CREATED` state will be very short-lived for normal case, it will change to `INIT` very quickly(<1s), so no need to show this state on frontend.